File: dfa_scraper.py
"""
DFA (Department of Foreign Affairs) News Scraper - Python version

Uses Wayback Machine archives to access dfa.gov.ph articles about WPS/SCS.
Direct access is blocked by Cloudflare, but Wayback Machine archives work.

Usage: python3 dfa_scraper.py
Output: dfa_WPS_data.json
"""

import json
import os
import re
import time
import requests
from datetime import datetime
from collections import defaultdict

# Configuration
CONFIG = {
    'dateFrom': '2022-07-01',
    'dateTo': '2026-05-31',
    'outputFile': os.path.join(os.path.dirname(__file__), 'dfa_WPS_data.json'),
    'delay': 1.0,  # seconds between Wayback Machine requests
}

# WPS/SCS keywords for URL filtering
URL_WPS_KEYWORDS = [
    'west-philippine', 'west-philippine', 'west_philippine',
    'south-china-sea', 'south+china+sea',
    '/wps', '-wps', 'wps.', 'wps-',
    'maritime', 'arbitral', 'unclos',
    'kalayaan', 'pag-asa', 'ayungin',
    'fisheries', 'coast-guard', 'coastguard',
    'reef', 'sovereignty', 'territorial',
    'chinese-vessel', 'china-sea',
    'naval', 'exclusive-economic', 'continental-shelf',
    'submission', 'extended-continental',
    'benham', 'philippine-rise',
    'scarborough', 'spratly', 'pancake',
]

# Primary content keywords that indicate WPS/SCS relevance
CONTENT_PRIMARY = [
    'west philippine sea', 'south china sea',
    'west philippine sea (wps)', 'wps',
]

# Secondary content keywords (weaker signal)
CONTENT_SECONDARY = [
    'arbitral award', 'unclos', 'ayungin shoal',
    'pagtindig para sa west philippine sea',
    'defending west philippine',
    'chinese coast guard', 'fishing ban',
    'bajo de masinloc', 'scarborough shoal',
    'kalayaan island', 'pag-asa island',
    'extended continental shelf',
    'maritime security',
]


def is_wps_url(url):
    """Check if URL slug suggests WPS/SCS content."""
    slug = url.lower()
    return any(kw in slug for kw in URL_WPS_KEYWORDS)


def has_wps_content(text):
    """Check if text contains WPS/SCS keywords."""
    lower = text.lower()
    primary_matches = [kw for kw in CONTENT_PRIMARY if kw in lower]
    secondary_matches = [kw for kw in CONTENT_SECONDARY if kw in lower]
    return primary_matches, secondary_matches


def extract_date(text):
    """Extract article date from text content."""
    # Look for "DD Month YYYY" pattern
    patterns = [
        r'(\d{1,2}\s+(?:January|February|March|April|May|June|July|August|September|October|November|December)\s+\d{4})',
        r'(\d{4}-\d{2}-\d{2})',
    ]
    for pat in patterns:
        match = re.search(pat, text, re.IGNORECASE)
        if match:
            try:
                if '-' in match.group(1):
                    return match.group(1)
                else:
                    dt = datetime.strptime(match.group(1), '%d %B %Y')
                    return dt.strftime('%Y-%m-%d')
            except:
                pass
    return ''


def fetch_wayback_article(url, timestamp):
    """Fetch an archived DFA article from Wayback Machine."""
    wb_url = f"https://web.archive.org/web/{timestamp}/{url}"

    headers = {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36'
    }

    try:
        r = requests.get(wb_url, timeout=30, headers=headers)
        if r.status_code != 200:
            return None

        html = r.text

        # Skip if blocked by Cloudflare
        if 'Just a moment' in html:
            return None

        # Extract title from HTML
        title_match = re.search(r'<title>(.*?)</title>', html, re.DOTALL)
        title = title_match.group(1).strip() if title_match else ''

        # Strip HTML tags to get text
        text = re.sub(r'<script[^>]*>.*?</script>', '', html, flags=re.DOTALL)
        text = re.sub(r'<style[^>]*>.*?</style>', '', text, flags=re.DOTALL)
        text = re.sub(r'<[^>]+>', ' ', text)
        # Clean up whitespace
        text = re.sub(r'\s+', ' ', text).strip()
        # Remove Wayback Machine specific text
        text = re.sub(r'The Wayback Machine - https://web\.archive\.org/.*?\s', '', text)
        text = re.sub(r'__wm\.\w+\([^)]+\)', '', text)

        # Check if we got meaningful content
        if len(text) < 100:
            return None

        # Extract date
        article_date = extract_date(text)

        return {
            'title': title,
            'text': text,
            'articleDate': article_date,
            'wbTimestamp': timestamp,
        }

    except Exception as e:
        return None


def load_urls(filepath):
    """Load URLs from a tab-separated file (timestamp\turl)."""
    urls = {}
    if os.path.exists(filepath):
        with open(filepath, 'r', encoding='utf-8') as f:
            for line in f:
                line = line.strip()
                if '\t' in line:
                    ts, url = line.split('\t', 1)
                    clean = url.split('?')[0].rstrip('/')
                    # Filter out listing pages
                    slug = clean.split('/')[-1] if '/' in clean else ''
                    if slug and len(slug) > 5 and not slug.startswith('start='):
                        if clean not in urls or ts > urls[clean]:
                            urls[clean] = ts
    return urls


def main():
    print('=== DFA News Scraper (via Wayback Machine) ===')
    print(f'Date range: {CONFIG["dateFrom"]} ~ {CONFIG["dateTo"]}')
    print()

    # Step 1: Load all URL sources
    script_dir = os.path.dirname(__file__)
    all_urls = {}

    for fname in ['dfa_article_urls.txt', 'dfa_statements_urls.txt']:
        fpath = os.path.join(script_dir, fname)
        urls = load_urls(fpath)
        print(f'Loaded {len(urls)} URLs from {fname}')
        for url, ts in urls.items():
            if url not in all_urls or ts > all_urls[url]:
                all_urls[url] = ts

    print(f'Total combined articles: {len(all_urls)}')

    # Step 2: Filter by URL keywords
    candidates = []
    for url, ts in all_urls.items():
        if is_wps_url(url):
            # Skip known false positive patterns
            slug = url.lower()
            if 'west-edmonton' in slug or 'western-australia' in slug or 'western-china' in slug:
                continue
            candidates.append({'url': url, 'timestamp': ts, 'slug': url.split('/')[-1]})

    print(f'URL-filtered candidates: {len(candidates)}')
    print()

    # Step 3: Fetch each candidate from Wayback Machine
    results = []
    total_wps = 0

    for i, cand in enumerate(candidates):
        slug_short = cand['slug'][:60]
        print(f'[{i+1}/{len(candidates)}] {slug_short}... ', end='', flush=True)

        data = fetch_wayback_article(cand['url'], cand['timestamp'])

        if data and len(data['text']) > 100:
            primary_kw, secondary_kw = has_wps_content(data['text'])
            if primary_kw or secondary_kw:
                results.append({
                    'title': data['title'].replace(' | Department of Foreign Affairs', '').replace(' - DFA', '').strip(),
                    'url': cand['url'],
                    'text': data['text'][:50000],
                    'date': data['articleDate'],
                    'wbTimestamp': data['wbTimestamp'],
                    'keywords': primary_kw + secondary_kw,
                    'source': 'DFA',
                    'country': 'Philippines',
                })
                total_wps += 1
                print(f'✓ WPS ({len(primary_kw + secondary_kw)} kw)')
            else:
                print(f'✗ no WPS')
        else:
            print(f'✗ empty/failed')

        time.sleep(CONFIG['delay'])

    # Step 4: Filter by date range
    from_date = datetime.fromisoformat(CONFIG['dateFrom'])
    to_date = datetime.fromisoformat(CONFIG['dateTo'])
    in_range = []

    for article in results:
        if article['date']:
            try:
                d = datetime.fromisoformat(article['date']) if 'T' in article['date'] else datetime.strptime(article['date'], '%Y-%m-%d')
                if from_date <= d <= to_date:
                    in_range.append(article)
            except:
                in_range.append(article)
        else:
            in_range.append(article)

    # Step 5: Save results
    print(f'\n=== Results ===')
    print(f'Total candidates: {len(candidates)}')
    print(f'With WPS content: {len(results)}')
    print(f'In date range: {len(in_range)}')

    output = {
        'config': CONFIG,
        'scrapedAt': datetime.now().isoformat(),
        'totalCandidates': len(candidates),
        'totalWithWps': len(results),
        'inRangeCount': len(in_range),
        'articles': in_range,
    }

    with open(CONFIG['outputFile'], 'w', encoding='utf-8') as f:
        json.dump(output, f, ensure_ascii=False, indent=2)

    print(f'\nSaved to {CONFIG["outputFile"]}')

    # Print year distribution
    year_dist = defaultdict(int)
    for a in in_range:
        d = a.get('date', '')
        if d and len(d) >= 4:
            year_dist[d[:4]] += 1
    print(f'Year distribution: {dict(sorted(year_dist.items()))}')

    # Print some sample titles
    print(f'\nSample articles:')
    for a in in_range[:10]:
        print(f'  {a["date"][:10] if a["date"] else "N/A"} - {a["title"][:70]}')


if __name__ == '__main__':
    main()

File: dfa_scraper.js
/**
 * DFA (Department of Foreign Affairs) News Scraper
 *
 * Uses Wayback Machine archives to access dfa.gov.ph articles about WPS/SCS
 * (direct access is blocked by Cloudflare)
 *
 * Usage: node dfa_scraper.js
 * Output: dfa_WPS_data.json
 */

const { chromium } = require('playwright');
const fs = require('fs');
const path = require('path');

const CONFIG = {
    dateFrom: '2022-07-01',
    dateTo: '2026-05-31',
    outputFile: path.join(__dirname, 'dfa_WPS_data.json'),
    minTextLength: 100, // Minimum text length to consider valid
};

// WPS/SCS keywords for URL filtering (first pass) and content filtering (second pass)
const URL_WPS_KEYWORDS = [
    'west-philippine', 'west+philippine', 'west_philippine',
    'south-china-sea', 'south+china+sea',
    '/wps', '-wps', 'wps-', 'wps.',
    'maritime', 'arbitral', 'unclos',
    'kalayaan', 'pag-asa', 'ayungin',
    'fisheries', 'coast-guard', 'coastguard',
    'navy', 'naval', 'reef', 'pacific',
    'sovereignty', 'territorial',
    'chinese-vessel', 'china-sea',
    'sea-security', 'east-sea',
];

const CONTENT_WPS_KEYWORDS = [
    'west philippine sea', 'south china sea', 'wps',
    'arbitral award', 'unclos',
];

function sleep(ms) {
    return new Promise(r => setTimeout(r, ms));
}

/**
 * Download and extract text from a Wayback Machine archived DFA article
 */
async function fetchWaybackArticle(browser, url, timestamp) {
    const context = await browser.newContext({
        userAgent: 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36'
    });
    const page = await context.newPage();

    // Use the most recent timestamp from our data
    const wbUrl = `https://web.archive.org/web/${timestamp}/${url}`;

    try {
        await page.goto(wbUrl, { waitUntil: 'domcontentloaded', timeout: 30000 });
        await sleep(1000);

        const title = await page.title();

        // Extract all text content
        const text = await page.evaluate(() => {
            const body = document.body;
            if (!body) return '';

            // Remove Wayback Machine navigation elements
            const waybackBanner = document.querySelector('#wm-ipp-base, #wm-ipp, .wm-ipp, #donato, #donation-description');
            if (waybackBanner) waybackBanner.remove();

            // Get the actual article content
            const contentSelectors = [
                '.item-page', '#content', '.content',
                'article', '.article', '.news-wrapper',
                '.main-content', '#main-content',
                '.tileItem', '.news-single',
            ];

            for (const sel of contentSelectors) {
                const el = document.querySelector(sel);
                if (el && el.textContent.trim().length > 200) {
                    return el.textContent;
                }
            }

            // Fallback: get body text
            return body.textContent || '';
        });

        // Clean up text
        const cleanText = text
            .replace(/The Wayback Machine - .*?\n/, '')
            .replace(/__wm\.\w+\([^)]+\)/g, '')
            .replace(/https:\/\/web\.archive\.org\/web\/[^/\s]+\//g, '')
            .replace(/\s+/g, ' ')
            .trim();

        // Extract date from text (look for "DD Month YYYY" pattern)
        const dateMatch = cleanText.match(/\b(\d{1,2}\s+(?:January|February|March|April|May|June|July|August|September|October|November|December)\s+\d{4})\b/i);

        // Also try ISO date
        const isoMatch = cleanText.match(/\b(\d{4}-\d{2}-\d{2})\b/);

        let articleDate = '';
        if (dateMatch) {
            try {
                const d = new Date(dateMatch[1]);
                if (!isNaN(d.getTime())) {
                    articleDate = d.toISOString();
                }
            } catch(e) {}
        }

        return {
            title: title.replace(/- Philippine Information Agency|- DFA|- Department of Foreign Affairs/i, '').trim(),
            text: cleanText,
            articleDate: articleDate,
            wbTimestamp: timestamp,
        };

    } catch (err) {
        return null;
    } finally {
        await page.close();
        await context.close();
    }
}

/**
 * Check if URL slug suggests WPS/SCS content
 */
function isWpsUrl(url) {
    const slug = url.toLowerCase();
    return URL_WPS_KEYWORDS.some(kw => slug.includes(kw));
}

/**
 * Check if text content contains WPS/SCS keywords
 */
function hasWpsContent(text) {
    const lower = text.toLowerCase();
    const matches = CONTENT_WPS_KEYWORDS.filter(kw => lower.includes(kw));
    return matches;
}

async function main() {
    console.log('=== DFA News Scraper (via Wayback Machine) ===');

    // Step 1: Load all DFA URLs from both sections
    const articleUrls = new Map(); // url -> latest timestamp

    // Load from the article URLs file
    if (fs.existsSync(path.join(__dirname, 'dfa_article_urls.txt'))) {
        const data = fs.readFileSync(path.join(__dirname, 'dfa_article_urls.txt'), 'utf-8');
        for (const line of data.trim().split('\n')) {
            const [ts, url] = line.split('\t');
            if (ts && url && ts >= '202207' && ts <= '202605') {
                const cleanUrl = url.split('?')[0].replace(/\/$/, '');
                // Only individual articles (not listing pages)
                if (cleanUrl.includes('postsupdate') && !cleanUrl.endsWith('postsupdate')) {
                    const slug = cleanUrl.split('/').pop();
                    if (slug && slug.length > 10) { // has a meaningful slug
                        if (!articleUrls.has(cleanUrl) || ts > articleUrls.get(cleanUrl)) {
                            articleUrls.set(cleanUrl, ts);
                        }
                    }
                }
            }
        }
    }

    // Also load from statements section
    const statementsFile = path.join(__dirname, 'dfa_statements_urls.txt');
    if (fs.existsSync(statementsFile)) {
        const data = fs.readFileSync(statementsFile, 'utf-8');
        for (const line of data.trim().split('\n')) {
            const [ts, url] = line.split('\t');
            if (ts && url) {
                const cleanUrl = url.split('?')[0].replace(/\/$/, '');
                if (!articleUrls.has(cleanUrl) || ts > articleUrls.get(cleanUrl)) {
                    articleUrls.set(cleanUrl, ts);
                }
            }
        }
    }

    console.log(`\nTotal articles collected: ${articleUrls.size}`);

    // Step 2: Filter by URL keywords
    const candidates = [];
    for (const [url, ts] of articleUrls) {
        if (isWpsUrl(url)) {
            const slug = url.split('/').pop();
            candidates.push({ url, ts, slug });
        }
    }

    console.log(`URL-filtered candidates (WPS keywords in URL): ${candidates.length}`);

    // Step 3: Fetch each candidate from Wayback Machine
    const browser = await chromium.launch({
        headless: true,
        args: ['--no-sandbox', '--disable-blink-features=AutomationControlled']
    });

    const results = [];
    try {
        for (let i = 0; i < candidates.length; i++) {
            const { url, ts, slug } = candidates[i];
            process.stdout.write(`[${i+1}/${candidates.length}] ${slug.substring(0, 50)}... `);

            const data = await fetchWaybackArticle(browser, url, ts);
            if (data && data.text.length > CONFIG.minTextLength) {
                const kwMatches = hasWpsContent(data.text);
                if (kwMatches.length > 0) {
                    results.push({
                        title: data.title,
                        url: url,
                        text: data.text.substring(0, 50000),
                        date: data.articleDate,
                        wbTimestamp: data.wbTimestamp,
                        keywords: kwMatches,
                        source: 'DFA',
                        country: 'Philippines',
                    });
                    console.log(`✓ (${data.title.substring(0, 40)})`);
                } else {
                    console.log(`✗ (no WPS content)`);
                }
            } else {
                console.log(`✗ (empty/failed)`);
            }

            // Be polite to Wayback Machine
            await sleep(2000);
        }

        // Step 4: Filter by date range
        const fromDate = new Date(CONFIG.dateFrom);
        const toDate = new Date(CONFIG.dateTo);
        const inRange = results.filter(a => {
            if (!a.date) return true; // include if no date found
            const d = new Date(a.date);
            return d >= fromDate && d <= toDate;
        });

        console.log(`\n=== Results ===`);
        console.log(`Total fetched: ${candidates.length}`);
        console.log(`With WPS content: ${results.length}`);
        console.log(`In date range: ${inRange.length}`);

        // Step 5: Save
        const output = {
            config: CONFIG,
            scrapedAt: new Date().toISOString(),
            totalCandidates: candidates.length,
            totalArticles: results.length,
            inRangeCount: inRange.length,
            articles: inRange,
        };

        fs.writeFileSync(CONFIG.outputFile, JSON.stringify(output, null, 2), 'utf-8');
        console.log(`\nSaved to ${CONFIG.outputFile}`);

    } catch (err) {
        console.error('Fatal:', err.message);
    } finally {
        await browser.close();
    }
}

// Main
main().catch(console.error);
